include/envoy/http/filter.h

```diff
@@ -396,6 +405,14 @@ class StreamDecoderFilter : public StreamFilterBase {
    */
   virtual FilterTrailersStatus decodeTrailers(HeaderMap& trailers) PURE;
 
+  /**
+   * Called with decoded metadata. Add new metadata to metadata_map directly. Do not call
+   * StreamDecoderFilterCallbacks::addDecodedMetadata() to add new metadata.
+   *
+   * @param metadata supplies the decoded metadata.
+   */
+  virtual FilterMetadataStatus decodeMetadata(MetadataMap& metadata_map) PURE;
```

Comment:
Thanks for the feedback!

Actually, because metadata is generally a hop-by-hop message, there is no need to wait for data or events to happen. All filters should return FilterMetadataStatus::Continue no matter they process metadata or not. As a result, metadata's process is kind of independent. When metadata is received or generated, it goes through all the filters's decodeMetadata() immediately, and then send to the next hop right away if the connection pool is ready. If the connection is not ready, metadata is stored in the router filter and wait for the connection to be ready. No further process of the metadata is needed. In summery, it doesn't interact with the process of other frame types, and is not part of the continueDecoding/commonContinue logic.

Sorry I am not familiar with ext_authz filter. Could you please elaborate on the problems that can be caused by metadata? Thanks!

Reply:
If `HEADER, METADATA, DATA` is a possible sequence then I do think blindly returning `FilterMetadataStatus::Continue` will be a problem.

Filters like `ext_authz` (as well as isito mixer filter) does call external service to check if the request can be passed to upstream. It initiates a RPC call (REST or gRPC) to an external service and `StopIteration` until the external service respond (or timeout, fail), so in case of `HEADER, METADATA, DATA` sequence, decodeMetadata has to return `StopIteration` as well during the wait for external service. See [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/ext_authz/ext_authz.cc#L75) for what it does for `decodeData` and `decodeTrailers`.

Blindly returning `Continue` will result if `METADATA` arrives before external service respond/timeout, it bypass the result of that and being proxied to upstream.

Reply:
Yeah agree this is a problem and I'm not sure the best way of handling this, because very few filters are going to care about metadata. This will definitely require filters that do the stop/continue logic that @lizan talks about to be modified. I hate to say this, but I'm wondering if we should make metadata callbacks opt-in somehow and not a primary callback interface. If a filter does not opt-in to metadata, the HCM would deal with the correct stop/continue logic automatically. Thoughts?

Reply:
What about a more general model for L7 events, where filters can opt-in to arbitrary events? The existing subclass/inheritance model works great for a few fixed events, but it seems that as we scale we don't want to force breaking API changes on all filters (plus, many don't care about most HTTP pipeline events beyond headers/data/trailers).

This could be done with dynamic registration of callbacks or handler interface implementations at construction time.

Reply:
Hi @lizan @alyssawilk and @mattklein123 , since we now have the StopAll status for headers ([PR](https://github.com/envoyproxy/envoy/pull/5954)), I am trying to apply it to the existing filters, starting with ext_authz. (Please let me know if we want metadata to use StopAll first, and then existing filters).

I have two questions:
1. Is there a procedure envoy follows to make production changes? If the change breaks production, is there a way to roll back?
2. The second question is about ext_authz filter. Maybe @lizan knows the answer. We can config ext_authz to send auth check after the full request body is received ([here](https://github.com/envoyproxy/envoy/blob/fa69fad0de6b63a254ce6a9f8164b31163a1ada0/source/extensions/filters/http/ext_authz/ext_authz.cc#L92)). But looks like auth check only needs request headers ([here](https://github.com/envoyproxy/envoy/blob/fa69fad0de6b63a254ce6a9f8164b31163a1ada0/source/extensions/filters/http/ext_authz/ext_authz.cc#L25)). I think it could be a way to buffer body when StopAll option is not available or it could be a feature needed by someone. Do you know if we should keep it with the new StopAlll option?

Thanks!

Reply:
I can answer (1) which is we definitely revert PRs if they cause production issues so don't worry about that.

Reply:
> But looks like auth check only needs request headers ([here](https://github.com/envoyproxy/envoy/blob/fa69fad0de6b63a254ce6a9f8164b31163a1ada0/source/extensions/filters/http/ext_authz/ext_authz.cc#L25)).

@soya3129 No it needs decodingBuffer(), which is retrieved via callback here: https://github.com/envoyproxy/envoy/blob/075edf802b9ef94bdf26fddc9e775f8bb89df92d/source/extensions/filters/common/ext_authz/check_request_utils.cc#L121

Reply:
Ah, got it. Will preserve data buffering. Thanks!!

Reply:
With Matt's permission I'd lean towards making the base class implementation return Continue, and landing these and tests in a follow up PR.  I did a recent change like that for the rc details and I think it is helpful to allow folks to fix their own filter implementations.

Alternately for this one we might want to keep the base class impl around

Either way I think we may have to move some of the existing Envoy filters to StopAllIteration before they work well with metadata.

Reply:
Added a todo to remind myself.

@alyssawilk, which filters would you recommend to check StopAllIteration? I can try to make the change. Thanks!

Reply:
> With Matt's permission I'd lean towards making the base class implementation return Continue, and landing these and tests in a follow up PR. I did a recent change like that for the rc details and I think it is helpful to allow folks to fix their own filter implementations.

Agreed. I think so few filters are going to care about metadata that we should try to handle this transparently. IMO having the base class return continue until we figure out a full solution makes sense to me.

Reply:
Added default return. Thanks!